### PR TITLE
[Core] Use static callback instead of dynamic notification listener

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

Currently, we use a notification listener for receiving object notifications from plasma store. However, this method has several flaws:

1. The subscription of the listener is dynamic, so we need to add locks to ensure the subscribers won't change while we are publishing notifications.
2. This makes debugging harder.
3. We need to go through all the objects to pick up sealed objects when there is a new subscriber so that we won't lose previous sent messages. In the worst case, this may block plasma store & raylet for a long time.

This new design replaces the old notification subscriber with static callbacks. It also replaces some outdated data structure.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -167,13 +167,40 @@ void HeartbeatSender::Heartbeat() {
...
+                         const NodeManagerConfig &config,
+                         const ObjectManagerConfig &object_manager_config,
                          std::shared_ptr<gcs::GcsClient> gcs_client,
-                         std::shared_ptr<ObjectDirectoryInterface> object_directory,
-                         std::function<bool(const ObjectID &)> is_plasma_object_spillable)
+                         std::shared_ptr<ObjectDirectoryInterface> object_directory)
     : self_node_id_(self_node_id),
       io_service_(io_service),
-      object_manager_(object_manager),
+      object_manager_(
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:176]: Nice!

File: src/ray/object_manager/object_manager.cc [hunk 2]
```diff
@@ -54,12 +55,35 @@ ObjectManager::ObjectManager(
...
+    std::function<void()> object_store_full_callback,
+    AddObjectCallback add_object_callback, DeleteObjectCallback delete_object_callback)
     : main_service_(&main_service),
       self_node_id_(self_node_id),
       config_(config),
       object_directory_(std::move(object_directory)),
-      object_store_internal_(config, spill_objects_callback, object_store_full_callback),
+      object_store_internal_(config, spill_objects_callback, object_store_full_callback,
+                             /*add_object_callback=*/
+                             [=](const ObjectInfo &object_info) {
```
Comment (rkooo567) [inline src/ray/object_manager/object_manager.cc:66]: Isn't `[this, add_object_callback=std::move(add_object_callback)]` enough?

**Review by rkooo567 — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Comment by suquark** [comment]
I need to fix the windows build before merging it:

```
(22:46:43) ERROR: BUILD.bazel:279:11: Couldn't build file _objs/plasma_store_server_lib/store.obj: C++ compilation of rule '//:plasma_store_server_lib' failed (Exit 2): cl.exe failed: error executing command 
```

**Comment by suquark** [comment]
windows build is fixed now

**Pull request merged**